File: 11.container-with-most-water.go
package leetcode

func maxArea(height []int) int {
	if len(height) < 2 {
		return 0
	}
	var area int
	// 双指针，每次移动较小的指针
	// 因此面积取决于较小的高度，如果移动较大的指针，面积也不可能超过当前的值
	for l, r := 0, len(height)-1; l < r; {
		h := min(height[l], height[r])
		w := r - l
		area = max(area, h*w)
		if height[l] > height[r] {
			r--
		} else {
			l++
		}
	}
	return area
}

File: pointers/11.container-with-most-water.py
#
# @lc app=leetcode id=11 lang=python3
#
# [11] Container With Most Water
#


# @lc code=start
class Solution:
    def maxArea(self, height) -> int:
        mArea = 0
        l = 0
        r = len(height) - 1
        while (l < r):
            area = min(height[l], height[r]) * (r - l)
            mArea = max(mArea, area)
            if height[l] <= height[r]:
                l += 1
            else:
                r -= 1
        return mArea


# @lc code=end
s = Solution()
s.maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7])
